### PR TITLE
Extend and update oneAPI support

### PR DESCRIPTION
**Pull request by fwyzard**
Update the code structure splitting the main and analyzer parts. 
Update to the API and features available in oneAPI beta 4.
Add support for building with the Intel and Codeplay open source implementations.
Run over all available devices, or those selected based on the command line argument.
Use an in-order queue to serialise the memory operations and kernel launch.

### CODE REVIEW

**Comment by fwyzard** [comment]
Sample output from `patatrack01`:
```
$ ./test-oneapi 
Got 53376 for cabling, wordCounter 36328

Running on SYCL device Intel(R) FPGA Emulation Device, driver version 2020.9.1.0
work groups: 9, work items per group: 4096
Output: 1699 modules on device
Output: 0 modules in 13.1855 us

Running on SYCL device Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz, driver version 2020.9.1.0.18
work groups: 9, work items per group: 4096
Output: 1699 modules on device
Output: 1 modules in 23.5011 us

Running on SYCL device Intel(R) Gen9 HD Graphics NEO, driver version 20.05.15524
work groups: 24, work items per group: 256
Output: 1699 modules on device
Output: 1699 modules in 14.3854 us

Running on SYCL device SYCL host device, driver version 1.2
work groups: 16, work items per group: 1
Output: 1699 modules on device
Output: 1699 modules in 20.7043 us
```

Note the wrong results using the CPU or FPGA emulator backends; the behaviour has already been reported to Intel.

File: analyzer_oneapi.cc [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  void analyze(cl::sycl::device device, Input const &input, Output &output, double &totaltime) {
+#if __SYCL_COMPILER_VERSION <= 20200118
+    // Intel oneAPI beta 4
+    cl::sycl::ordered_queue queue{device, exception_handler};
+#else
+    // Intel SYCL branch
+    cl::sycl::queue queue{device, exception_handler, cl::sycl::property::queue::in_order()};
+#endif
```
Comment (makortel) [inline analyzer_oneapi.cc:36]: Do I understand correctly that the upper branch is essentially  for the Codeplay's implementatoin, and the lower branch is for Intel's (which is more recent than the "beta 4")?

**Review by makortel — APPROVED** [review]

File: analyzer_oneapi.cc (hunk 1 above)
Comment (fwyzard) [reply analyzer_oneapi.cc:36]: No, they are both for the Intel branch:
  - the top part is for the [beta04](https://software.intel.com/en-us/forums/intel-oneapi-base-toolkit/topic/848563) compiler, that was branched off some time in January
  - the bottom branch is for the HEAD of the [sycl](https://github.com/intel/llvm/tree/sycl) branch

The Codeplay branch is slightly less recent, but still able to work with the bottom branch.
Once beta05 is available (within a couple of months) I can remove the check.

Comment (makortel) [reply analyzer_oneapi.cc:36]: Ok. But now I'm confused: are the comments in the code correct, or your GH comment above ("top = HEAD, bottom = beta04")?

Comment (fwyzard) [reply analyzer_oneapi.cc:36]: Fixed.

Comment (makortel) [reply analyzer_oneapi.cc:36]: Thanks!

**Pull request merged**